Problems during s2i build on upgrading kubernetes-client to latest version

**Please consider this whenever upgrading fabric8-maven-plugin's kubernetes-client dependency; Right now fmp uses a pretty old version of kubernetes client 3.0.3.**

I was trying to test something using the latest SNAPSHOT version of kubernetes-client on fabric8 maven plugin. But after configuring fmp to use latest client, I saw this while trying to deploy a simple application in samples/spring-boot:
```
[INFO] Copying files to /home/rohan/work/repos/fabric8-maven-plugin/samples/spring-boot/target/docker/fabric8-maven-sample-spring-boot/latest/build/maven
[INFO] Building tar: /home/rohan/work/repos/fabric8-maven-plugin/samples/spring-boot/target/docker/fabric8-maven-sample-spring-boot/latest/tmp/docker-build.tar
[INFO] F8: [fabric8-maven-sample-spring-boot:latest] "spring-boot": Created docker source tar /home/rohan/work/repos/fabric8-maven-plugin/samples/spring-boot/target/docker/fabric8-maven-sample-spring-boot/latest/tmp/docker-build.tar
[INFO] F8: Using BuildServiceConfig fabric8-maven-sample-spring-boot-s2i for Source strategy
[INFO] F8: Creating ImageStream fabric8-maven-sample-spring-boot
[INFO] F8: Starting Build fabric8-maven-sample-spring-boot-s2i
[INFO] F8: Waiting for build fabric8-maven-sample-spring-boot-s2i-34 to complete...
[INFO] F8: ==================================================================
[INFO] F8: Starting S2I Java Build .....
[INFO] F8: S2I binary build from fabric8-maven-plugin detected
[INFO] F8: Copying binaries from /tmp/src/maven to /deployments ...
[INFO] F8: Checking for SpringBoot archive...
[INFO] F8: Found fabric8-maven-sample-spring-boot-3.5-SNAPSHOT.jar...
[INFO] F8: ... done
[WARNING] Ignoring error for already closed/closing connection
[ERROR] F8: Failed to execute the build [Unable to build the image using the OpenShift build service]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 19.026 s
[INFO] Finished at: 2018-04-04T16:22:15+05:30
[INFO] Final Memory: 52M/471M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.fabric8:fabric8-maven-plugin:3.5-SNAPSHOT:build (default) on project fabric8-maven-sample-spring-boot: Failed to execute the build: Unable to build the image using the OpenShift build service: NullPointerException -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,7 @@ After this we will switch probably to real [Semantic Versioning 2.0.0](http://se
 * Fix 1185: K8s: resource fragment containing compute resources for containers triggers a WARNING.
 * Fix 1237: When trimImageInContainerSpec is enabled, the generated yaml is incorrect
 * Fix 1245: Use released version of Booster in place of master always in regression test
+* Fix 1263: Display a warning in case of premature close of the build watcher by kubernetes client
 
 ###3.5.38
 * Feature 1209: Added flag fabric8.openshift.generateRoute which if set to false will not generate route.yml and also will not add Route resource in openshift.yml. If set to true or not set, it will generate rou  te.yml and also add Route resource in openshift.yml. By default its value is true.

--- a/core/src/main/java/io/fabric8/maven/core/service/openshift/OpenshiftBuildService.java
+++ b/core/src/main/java/io/fabric8/maven/core/service/openshift/OpenshiftBuildService.java
@@ -389,8 +389,7 @@ public class OpenshiftBuildService implements BuildService {
             try (Watch watcher = client.builds().withName(buildName).watch(buildWatcher)) {
                 // Check if the build is already finished to avoid waiting indefinitely
                 Build lastBuild = client.builds().withName(buildName).get();
-                String lastStatus = KubernetesResourceUtil.getBuildStatusPhase(lastBuild);
-                if (Builds.isFinished(lastStatus)) {
+                if (Builds.isFinished(KubernetesResourceUtil.getBuildStatusPhase(lastBuild))) {
                     log.debug("Build %s is already finished", buildName);
                     buildHolder.set(lastBuild);
                     latch.countDown();
@@ -398,12 +397,22 @@ public class OpenshiftBuildService implements BuildService {
 
                 waitUntilBuildFinished(latch);
                 logTerminateLatch.countDown();
+
                 build = buildHolder.get();
+                if (build == null) {
+                    log.debug("Build watcher on %s was closed prematurely", buildName);
+                    build = client.builds().withName(buildName).get();
+                }
                 String status = KubernetesResourceUtil.getBuildStatusPhase(build);
                 if (Builds.isFailed(status) || Builds.isCancelled(status)) {
-                    throw new MojoExecutionException("OpenShift Build " + buildName + ": " + KubernetesResourceUtil.getBuildStatusReason(build));
+                    throw new MojoExecutionException("OpenShift Build " + buildName + " error: " + KubernetesResourceUtil.getBuildStatusReason(build));
                 }
-                log.info("Build " + buildName + " " + status);
+
+                if (!Builds.isFinished(status)) {
+                    log.warn("Could not wait for the completion of build %s. It may be  may be still running (status=%s)", buildName, status);
+                } else {
+                    log.info("Build %s in status %s", buildName, status);
+                }
             }
         }
     }

--- a/core/src/main/java/io/fabric8/maven/core/util/KubernetesResourceUtil.java
+++ b/core/src/main/java/io/fabric8/maven/core/util/KubernetesResourceUtil.java
@@ -553,19 +553,16 @@ public class KubernetesResourceUtil {
     }
 
     public static String getBuildStatusPhase(Build build) {
-        String status = null;
-        BuildStatus buildStatus = build.getStatus();
-        if (buildStatus != null) {
-            status = buildStatus.getPhase();
+        if (build != null && build.getStatus() != null) {
+            return build.getStatus().getPhase();
         }
-        return status;
+        return null;
     }
 
     public static String getBuildStatusReason(Build build) {
-        BuildStatus buildStatus = build.getStatus();
-        if (buildStatus != null) {
-            String reason = buildStatus.getReason();
-            String phase = buildStatus.getPhase();
+        if (build != null && build.getStatus() != null) {
+            String reason = build.getStatus().getReason();
+            String phase = build.getStatus().getPhase();
             if (Strings.isNotBlank(phase)) {
                 if (Strings.isNotBlank(reason)) {
                     return phase + ": " + reason;

